src/app/fdctl/main1.c

```diff
@@ -20,17 +29,105 @@ struct action_alias ALIASES[] = {
+                     config_t * config ) {
+  uchar * bytes = mmap( NULL, sizeof( config_t ), PROT_READ, MAP_PRIVATE, config_fd, 0 );
+  if( FD_UNLIKELY( bytes == MAP_FAILED ) ) FD_LOG_RAW_ERR(( "mmap() failed (%i-%s)", errno, fd_io_strerror( errno ) ));
+  fd_memcpy( config, bytes, sizeof( config_t ) );
+  if( FD_UNLIKELY( munmap( bytes, sizeof( config_t ) ) ) ) FD_LOG_RAW_ERR(( "munmap() failed (%i-%s)", errno, fd_io_strerror( errno ) ));
+  if( FD_UNLIKELY( close( config_fd ) ) ) FD_LOG_RAW_ERR(( "close() failed (%i-%s)", errno, fd_io_strerror( errno ) ));
+}
+
+static int *
+map_log_memfd( int log_memfd ) {
+  void * shmem = mmap( NULL, 4096, PROT_READ | PROT_WRITE, MAP_SHARED, log_memfd, (off_t)0 );
+  if( FD_UNLIKELY( shmem==MAP_FAILED ) ) {
+    FD_LOG_RAW_ERR(( "mmap(NULL,sizeof(int),PROT_READ|PROT_WRITE,MAP_SHARED,memfd,(off_t)0) (%i-%s); ", errno, fd_io_strerror( errno ) ));
+  } else {
+    if( FD_UNLIKELY( mlock( shmem, 4096 ) ) ) {
+      FD_LOG_RAW_ERR(( "mlock(%p,4096) (%i-%s); unable to lock log file shared lock in memory\n",
+                       shmem, errno, fd_io_strerror( errno ) ));
+    }
+  }
+  return shmem;
+}
+
+/* Try to allocate an anonymous page of memory in a file descriptor
+   (memfd) for fd_log_private_shared_lock such that the log can strictly
+   sequence messages written by clones of the caller made after the
+   caller has finished booting the log.  Must be a file descriptor so
+   we can pass it through `execve` calls. */
+static int
+init_log_memfd( void ) {
+  int memfd = memfd_create( "fd_log_lock_page", 0U );
+  if( FD_UNLIKELY( -1==memfd) ) FD_LOG_RAW_ERR(( "memfd_create(\"fd_log_lock_page\",0) failed (%i-%s)", errno, fd_io_strerror( errno ) ));
+  if( FD_UNLIKELY( -1==ftruncate( memfd, 4096 ) ) ) {
+    FD_LOG_RAW_ERR(( "ftruncate(memfd,4096) failed (%i-%s)", errno, fd_io_strerror( errno ) ));
+  }
+  return memfd;
+}
+
+config_t
+fdctl_boot( int *        pargc,
+            char ***     pargv,
+            char const * log_path ) {
+  int config_fd = fd_env_strip_cmdline_int( pargc, pargv, "--config-fd", NULL, -1 );
+
+  config_t config = {0};
+  if( FD_UNLIKELY( config_fd >= 0 ) ) {
+    copy_config_from_fd( config_fd, &config );
+  } else {
+    config_parse( pargc, pargv, &config );
+    config.log.lock_fd = init_log_memfd();
+    config.log.log_fd  = -1;
+    if( FD_UNLIKELY( log_path ) )
+      strncpy( config.log.path, log_path, sizeof( config.log.path ) - 1 );
+  }
+
+  int * log_lock = map_log_memfd( config.log.lock_fd );
+  int pid = getpid1(); /* Need to read /proc since we might be in a PID namespace now */;
+
+  log_path = config.log.path;
+  if( FD_LIKELY( config.log.path[ 0 ]=='\0' ) ) log_path = NULL;
+
+  fd_log_private_boot1( log_lock,
+                        0UL,
+                        config.name,
+                        0UL,    /* Thread ID will be initialized later */
+                        "main", /* Thread will be initialized later */
```

Comment:
Maybe name this something expressing that it's uninitialized if it will be initialized later? Just like an empty string or "unset" or something